packages/trace-viewer/src/ui/snapshotTab.tsx

```diff
@@ -325,18 +325,59 @@ export type SnapshotUrls = {
   popoutUrl: string;
 };
 
+function isTestStep(action: ActionTraceEvent) {
+  return context(action).origin === 'testRunner' && action.callId.startsWith('test.step');
+}
+
 export function collectSnapshots(action: ActionTraceEvent | undefined): Snapshots {
   if (!action)
     return {};
 
   // if the action has no beforeSnapshot, use the last available afterSnapshot.
   let beforeSnapshot: Snapshot | undefined = action.beforeSnapshot ? { action, snapshotName: action.beforeSnapshot } : undefined;
-  let a = action;
-  while (!beforeSnapshot && a) {
-    a = prevInList(a);
-    beforeSnapshot = a?.afterSnapshot ? { action: a, snapshotName: a?.afterSnapshot } : undefined;
+  if (!beforeSnapshot) {
+    for (let a = prevInList(action); a; a = prevInList(a)) {
+      if (a.endTime <= action.startTime && a.afterSnapshot) {
+        beforeSnapshot = { action: a, snapshotName: a.afterSnapshot };
+        break;
+      }
+    }
   }
-  const afterSnapshot: Snapshot | undefined = action.afterSnapshot ? { action, snapshotName: action.afterSnapshot } : beforeSnapshot;
+  // For test.step, we can use snapshot from the first nested action.
+  if (!beforeSnapshot && isTestStep(action)) {
```

Comment:
- Why aren't we using the action tree here instead of relying on timing?
- Why only for "test steps" instead of any actions that have children but do not have snapshots?
- Why do we prefer "after snapshot from the previous action" instead of "before snapshot of the first child"?

Reply:
Updated as discussed offline.